### ISSUE TITLE
Memory leak with mirage-net-solo5(?) under load

### ISSUE THREAD

**Issue by mato** [Solo5/solo5#58]
Reporting this here as mirage-net-solo5 does not have issues enabled, and it's unclear if that is the root cause.

When running load tests against `static_website` using `siege -c 2 -b -t <TIME> http://10.0.0.2:8080/`, heap use grows rapidly until memory runs out.

Investigation so far:
- The unikernel survives a maximum siege of about `5s` before the ukvm memory allocation of 512MB runs out.
- The interaction of `io-page` with GC is unclear. `mirage-net-solo5` does have the tail-recursion fixes from `mirage-net-unix` in it, but these do not seem to help.
- Solo5 does not currently require page aligned buffers for I/O, so @hannesm made a quick test version of io-page to use `Cstruct` directly rather than the out-of-band page allocator. This can be OPAM-pinned with `opam pin io-page  git://github.com/hannesm/io-page#cstruct`.
- I added a call to `Gc.full_major ()` followed by logging of GC stats to `static_website` every two seconds to attempt to diagnose the GC behaviour.
- There is a separate bug (to be reported) where the unikernel appears to freeze (no GC logging output) after `siege` completes, and does not wake up until receiving at least one network packet.

Example output with GC stats and two `5s` siege runs -- GC values are `live_words` `minor_collections` `major_collections`:

```
2016-07-05 11:18.49: INF [gc] GC: 86061 2 2
2016-07-05 11:18.51: INF [gc] GC: 86068 3 4
2016-07-05 11:18.53: INF [gc] GC: 86068 4 6
2016-07-05 11:18.55: INF [gc] GC: 86068 5 8
2016-07-05 11:18.57: INF [gc] GC: 86068 6 10
Manager: configuration done
ARP: transmitting probe -> 10.0.0.1
ARP: updating 10.0.0.1 -> 0a:3f:94:7c:a1:51
2016-07-05 11:18.59: INF [gc] GC: 3345761 164 21
2016-07-05 11:19.01: INF [gc] GC: 6932473 444 26
//// (freeze until ping)
2016-07-05 11:19.13: INF [gc] GC: 7113271 555 28
2016-07-05 11:19.15: INF [gc] GC: 1038586 567 30
2016-07-05 11:19.17: INF [gc] GC: 93341 568 32
2016-07-05 11:19.19: INF [gc] GC: 93341 569 34
2016-07-05 11:19.21: INF [gc] GC: 93341 570 36
2016-07-05 11:19.23: INF [gc] GC: 1111774 620 43
2016-07-05 11:19.25: INF [gc] GC: 8258875 975 51
2016-07-05 11:19.27: INF [gc] GC: 8077803 1253 56
//// (freeze until ping)
2016-07-05 11:19.39: INF [gc] GC: 6400794 1308 58
2016-07-05 11:19.41: INF [gc] GC: 209872 1321 60
2016-07-05 11:19.43: INF [gc] GC: 97437 1322 62
2016-07-05 11:19.45: INF [gc] GC: 97437 1323 64
ARP: timeout 10.0.0.1
2016-07-05 11:19.47: INF [gc] GC: 97427 1324 66
2016-07-05 11:19.49: INF [gc] GC: 97427 1325 68
^CReceived SIGINT. Exiting
```

**Comment by mato** [comment]
Patch to add GC stats to `static_website`
[gc-stats-patch.txt](https://github.com/djwillia/solo5/files/347659/gc-stats-patch.txt)

**Comment by hannesm** [comment]
as mentioned on IRC, if we can reproduce this on a Unix system, the new memory profiler spacetime (https://github.com/ocaml/ocaml/pull/585) might help to track down the issue... not sure what is involved to port spacetime to solo5&xen (it is worth investigating), likely having some shared memory between host&guest where the data is dumped (similar to the tracing stuff)!?

**Comment by djwillia** [comment]
Is this for both `virtio` and `ukvm` targets?

**Comment by mato** [comment]
I haven't tested yet with `virtio`, but it shouldn't make a difference -- this is a Ocaml GC issue, not a Solo5 malloc issue.

**Comment by mato** [comment]
Just tested behaviour on `virtio` and it's identical except for the absence of the "freeze until next ping" after a `siege` run. So there is a separate ukvm-specific bug lurking somewhere which needs tracking down.

**Comment by mato** [comment]
I've fixed the ukvm-specific "freezes", this was due to a bug in `solo5_poll()` (#60).

**Comment by mato** [comment]
More observations:
- Rebuilt everything with OCaml 4.03.0 (testing until now was on 4.02.3). Running on a faster CPU (Core i7-5557U) and `static_website` keeps up with a `siege` load test indefinitely. Memory use (based on `ukvm` RSS) is at around 400MB.
- Running with the above configuration but less guest RAM (256MB) results in eating all available memory and aborting.
- @avsm suggests looking into tuning the GC to deal with memory pressure ahead of time (before we start to run out).

**Comment by mato** [comment]
Testing on `ukvm`, with hannes' iopage#cstruct branch: Lowering the Gc `space_overhead` tunable from the default of `80` to `20` results in a stable fully-loaded system, with memory use of ~200MB (OCaml 4.02.3, Intel Core i5 laptop) and ~145MB (OCaml 4.03.0, Intel Core i7 NUC).

Conclusion The `siege` load test is stable as long as it has enough memory and GC has been tuned correctly. Having said that, memory usage in hundreds of megabytes for such a test is excessive, so we need to look into reducing/limiting memory allocation in the stack.

**Comment by hannesm** [comment]
Is this still leaking?  we should in some FAQ (likely somewhere on mirage.io) document the GC tweaking (maybe it applies to Xen as well?)...

**Comment by mato** [comment]
It's definitely not leaking any longer. Latest mirage-dev with OCaml 4.03.0 with no GC tuning survives a repeated test with `siege`, with an ukvm RSS of around 330MB. I have not tried 4.02.3 again, I suspect that would need manual tuning.

Yes, we should document, and/or make it easier to tune GC for the Mirage unikernel. That's a separate issue so I'm closing this one.
